src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -229,86 +249,101 @@ public void ApplyChangeToWorkspace(ProjectId projectId, Func<CodeAnalysis.Soluti
+    public async Task ApplyBatchChangeToWorkspaceMaybe_NoLockAsync(bool useAsync, Func<SolutionChangeAccumulator, ProjectUpdateState, ProjectUpdateState> mutation, Action<ProjectUpdateState>? onAfterUpdate)
     {
-        Contract.ThrowIfFalse(_gate.CurrentCount == 0);
+        // We need the data from the accumulator across the lambda callbacks to SetCurrentSolutionAsync, so declare
+        // it here. It will be assigned in `transformation:` below (which may happen multiple times if the
+        // transformation needs to rerun).  Once the transformation succeeds and is applied, the
+        // 'onBeforeUpdate/onAfterUpdate' callbacks will be called, and can use the last assigned value in
+        // `transformation`.
+        SolutionChangeAccumulator solutionChanges = null!;
+        ProjectUpdateState projectUpdateState = null!;
+
+        var (didUpdate, newSolution) = await Workspace.SetCurrentSolutionAsync(
+            useAsync,
+            transformation: oldSolution =>
+            {
+                solutionChanges = new SolutionChangeAccumulator(oldSolution);
 
-        if (!solutionChanges.HasChange)
-            return;
+                // Reset the project update state in case this is a retry (we don't want to save the results of the last attempt).
+                projectUpdateState = _projectUpdateState;
+                projectUpdateState = mutation(solutionChanges, projectUpdateState);
 
-        Workspace.SetCurrentSolution(
-            _ => solutionChanges.Solution,
-            solutionChanges.WorkspaceChangeKind,
-            solutionChanges.WorkspaceChangeProjectId,
-            solutionChanges.WorkspaceChangeDocumentId,
+                // Note: If the accumulator showed no changes it will return oldSolution.  This ensures that
+                // SetCurrentSolutionAsync bails out immediately and no further work is done.
+                return solutionChanges.Solution;
+            },
+            changeKind: (_, _) => (solutionChanges.WorkspaceChangeKind, solutionChanges.WorkspaceChangeProjectId, solutionChanges.WorkspaceChangeDocumentId),
             onBeforeUpdate: (_, _) =>
             {
                 // Clear out mutable state not associated with the solution snapshot (for example, which documents are
                 // currently open).
                 foreach (var documentId in solutionChanges.DocumentIdsRemoved)
                     Workspace.ClearDocumentData(documentId);
-            });
+            },
+            onAfterUpdate: (_, _) =>
+            {
+                // Now that the project update has actually applied, we can apply the results of it.
+                // For example saving the state and updating file watchers for added/removed references.
+                ApplyProjectUpdateState(projectUpdateState);
+                if (onAfterUpdate != null)
+                {
+                    onAfterUpdate(projectUpdateState);
+                }
```

Comment:
```suggestion
                    onAfterUpdate?.Invoke(projectUpdateState);
```

Note: i don't have a better suggestion, but trying to keep track of the chain of onAfterUpdates is... hard :)